### PR TITLE
Fixes for XML reporting

### LINKED ISSUE #1182: Wrong name and class name in XML report for parameterized tests
A test like this

```java
package com.example;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
public class SomeTest {
   @ParameterizedTest
   @ValueSource(strings = { "val0", "val1" })
   void testSomething(String someValue) {
   }
}
```

results in an XML report (`TEST-junit-jupiter.xml`) like this:

```xml
...
<testcase name="testSomething(String)" classname="testSomething(String)" time="0">...</testcase>
<testcase name="testSomething(String)" classname="testSomething(String)" time="0">...</testcase>
...
```

A similar test in JUnit 4 (`@RunWith(Parameterized.class)`), however, produces something like this:

```xml
...
<testcase name="testSomething[0]" classname="com.example.SomeTest" time="0.0"/>
<testcase name="testSomething[1]" classname="com.example.SomeTest" time="0.0"/>
...
```

My "problem" in particular or rather observation is that after migrating to JUnit 5 the number of tests reported in sonar dropped and the tests only using `@ParameterizedTest` no longer show up. Since as far as I know, sonar parses those XML reports, it's the only explanation I have for that drop. This for sure is not limited to sonar but all tools parsing those XML reports.

So it obviously has no high priority, at all but it would be great if you could fix this at some point.

In case it is relevant: I am using gradle. 

```groovy
buildscript {
  repositories {
    mavenCentral()
  }
  dependencies {
    classpath 'org.junit.platform:junit-platform-gradle-plugin:1.0.2'
  }
}
repositories {
  mavenCentral()
}

apply plugin: 'eclipse'
apply plugin: 'java'
apply plugin: 'org.junit.platform.gradle.plugin'

dependencies {
  testCompile(
    'org.junit.jupiter:junit-jupiter-api:5.0.2',
    'org.junit.jupiter:junit-jupiter-params:5.0.2'
  )
  testRuntime(
    'org.junit.jupiter:junit-jupiter-engine:5.0.2',
    'org.junit.platform:junit-platform-launcher:1.0.2'
  )
}
```

## Deliverables

- [ ] Ensure that attribute values (e.g., `name` and `classname`) in XML reports make sense for the following use cases.
  - [ ] repeated tests
  - [ ] parameterized tests
  - [ ] dynamic containers
  - [ ] dynamic tests

### PR DESCRIPTION
**Pull request by marcphilipp**
## Overview

Fixes #1182. This pull request provides fixes related to XML reports emitted by `ConsoleLauncher`, Gradle plugin, and Maven Surefire.

1. Output class name also for dynamic tests and test template invocations
1. Suffix names with invocation indexes for dynamic tests and test template invocations
1. Output method parameter types also for Maven Surefire but as `testMethod{Type}` instead of `testMethod(Type)` due to a limitation of Surefire

Here's a reference test:

```java
class JUnit5Tests {

  @Test
  void myNormalTest() {}

  @RepeatedTest(3)
  void myRepeatedTest() {}

  @ParameterizedTest
  @ValueSource(strings = {
    "hello",
    "world"
  })
  void myParametrizedTest(String s) {}

  @TestFactory
  Stream<DynamicNode> myTestFactory() {
    return Stream.of(dynamicContainer("foo", Stream.of(dynamicTest("bar", () -> {}))));
  }

  @Nested
  class MyNestedTest {

    @Test
    void myNormalTest() {}

    @RepeatedTest(3)
    void myRepeatedTest() {}

    @ParameterizedTest
    @ValueSource(strings = {
            "hello",
            "world"
    })
    void myParametrizedTest(String s) {}

    @TestFactory
    Stream<DynamicNode> myTestFactory() {
      return Stream.of(dynamicContainer("foo", Stream.of(dynamicTest("bar", () -> {}))));
    }
  }
}
```

Here is its output when executed using Maven:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://maven.apache.org/surefire/maven-surefire-plugin/xsd/surefire-test-report.xsd" name="com.example.project.JUnit5Tests" time="0.055" tests="14" errors="0" skipped="0" failures="0">
  <properties>...</properties>
  <testcase name="myRepeatedTest[1]" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myRepeatedTest[1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myRepeatedTest[2]" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myRepeatedTest[2]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myRepeatedTest[3]" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myRepeatedTest[3]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myParametrizedTest{String}[1]" classname="com.example.project.JUnit5Tests" time="0.02"/>
  <testcase name="myParametrizedTest{String}[1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myParametrizedTest{String}[2]" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myParametrizedTest{String}[2]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myNormalTest" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myNormalTest" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
  <testcase name="myTestFactory[1][1]" classname="com.example.project.JUnit5Tests" time="0"/>
  <testcase name="myTestFactory[1][1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0"/>
</testsuite>
```

And here's the output for Gradle:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="JUnit Jupiter" tests="14" skipped="0" failures="0" errors="0" time="0.045" timestamp="2018-01-09T20:21:42">
<properties>...</properties>
<testcase name="myNormalTest()" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[method:myNormalTest()]
display-name: myNormalTest()
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[1]" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-template:myRepeatedTest()]/[test-template-invocation:#1]
display-name: repetition 1 of 3
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[2]" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-template:myRepeatedTest()]/[test-template-invocation:#2]
display-name: repetition 2 of 3
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[3]" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-template:myRepeatedTest()]/[test-template-invocation:#3]
display-name: repetition 3 of 3
]]></system-out>
</testcase>
<testcase name="myParametrizedTest(String)[1]" classname="com.example.project.JUnit5Tests" time="0.013">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-template:myParametrizedTest(java.lang.String)]/[test-template-invocation:#1]
display-name: [1] hello
]]></system-out>
</testcase>
<testcase name="myParametrizedTest(String)[2]" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-template:myParametrizedTest(java.lang.String)]/[test-template-invocation:#2]
display-name: [2] world
]]></system-out>
</testcase>
<testcase name="myTestFactory()[1][1]" classname="com.example.project.JUnit5Tests" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[test-factory:myTestFactory()]/[dynamic-container:#1]/[dynamic-test:#1]
display-name: bar
]]></system-out>
</testcase>
<testcase name="myNormalTest()" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[method:myNormalTest()]
display-name: myNormalTest()
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-template:myRepeatedTest()]/[test-template-invocation:#1]
display-name: repetition 1 of 3
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[2]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-template:myRepeatedTest()]/[test-template-invocation:#2]
display-name: repetition 2 of 3
]]></system-out>
</testcase>
<testcase name="myRepeatedTest()[3]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-template:myRepeatedTest()]/[test-template-invocation:#3]
display-name: repetition 3 of 3
]]></system-out>
</testcase>
<testcase name="myParametrizedTest(String)[1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-template:myParametrizedTest(java.lang.String)]/[test-template-invocation:#1]
display-name: [1] hello
]]></system-out>
</testcase>
<testcase name="myParametrizedTest(String)[2]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-template:myParametrizedTest(java.lang.String)]/[test-template-invocation:#2]
display-name: [2] world
]]></system-out>
</testcase>
<testcase name="myTestFactory()[1][1]" classname="com.example.project.JUnit5Tests$MyNestedTest" time="0">
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]/[class:com.example.project.JUnit5Tests]/[nested-class:MyNestedTest]/[test-factory:myTestFactory()]/[dynamic-container:#1]/[dynamic-test:#1]
display-name: bar
]]></system-out>
</testcase>
<system-out><![CDATA[
unique-id: [engine:junit-jupiter]
display-name: JUnit Jupiter
]]></system-out>
</testsuite>
```

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [X] There are no TODOs left in the code
- [X] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [X] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [X] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [X] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [X] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/RunListenerAdapter.java [hunk 1]
```diff
@@ -161,17 +162,21 @@ private SimpleReportEntry createReportEntry(TestIdentifier testIdentifier,
...
 	}
 
-	private String sourceLegacyReportingName(TestIdentifier testIdentifier) {
-		// @formatter:off
-		return testPlan.getParent(testIdentifier)
-				.map(TestIdentifier::getLegacyReportingName)
-				.orElse("<unrooted>");
-		// @formatter:on
+	private String getLegacyReportingName(TestIdentifier testIdentifier) {
+		// Surefire cuts of the name at the first '(' character. Thus, we have to pick a different
```
Comment (jbduncan) [inline junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/RunListenerAdapter.java]: "s/of/off"?

Comment (marcphilipp) [reply junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/RunListenerAdapter.java]: Yep, thanks!

Comment (jbduncan) [reply junit-platform-surefire-provider/src/main/java/org/junit/platform/surefire/provider/RunListenerAdapter.java]: You're very welcome! 😀

**Comment by sbrannen** [comment]
👏

**Pull request merged**